### ISSUE TITLE
[REVIEW]: udocker: a user oriented tool for unprivileged Linux containers

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#3295]
**Submitting author:** @jorge-lip (<a href="http://orcid.org/0000-0002-9142-2596">Jorge Gomes</a>)
**Repository:** <a href="https://github.com/indigo-dc/udocker" target ="_blank">https://github.com/indigo-dc/udocker</a>
**Version:** v1.1.7
**Editor:** @danielskatz
**Reviewers:** @vsoch
**Archive:** Pending

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

## Status

[![status](https://joss.theoj.org/papers/c14d991a006947bb45804eead0e830d0/status.svg)](https://joss.theoj.org/papers/c14d991a006947bb45804eead0e830d0)

Status badge code:

```
HTML: <a href="https://joss.theoj.org/papers/c14d991a006947bb45804eead0e830d0"><img src="https://joss.theoj.org/papers/c14d991a006947bb45804eead0e830d0/status.svg"></a>
Markdown: [![status](https://joss.theoj.org/papers/c14d991a006947bb45804eead0e830d0/status.svg)](https://joss.theoj.org/papers/c14d991a006947bb45804eead0e830d0)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/indigo-dc/udocker" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) by leaving comments in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@vsoch & @mviereck, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html. Any questions/concerns please let @danielskatz know.

✨ **Please start on your review when you are able, and be sure to complete your review in the next six weeks, at the very latest** ✨ 

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/indigo-dc/udocker">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@jorge-lip) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

### Conflict of interest

- [ ] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [ ] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [ ] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/indigo-dc/udocker">repository url</a>?
- [ ] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [ ] **Contribution and authorship:** Has the submitting author (@jorge-lip) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [ ] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [ ] **Installation:** Does installation proceed as outlined in the documentation?
- [ ] **Functionality:** Have the functional claims of the software been confirmed?
- [ ] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [ ] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [ ] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [ ] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [ ] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [ ] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [ ] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [ ] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [ ] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [ ] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [ ] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [ ] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks. @vsoch, @mviereck it looks like you're currently assigned to review this paper :tada:.

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

For example, to regenerate the paper pdf after making changes in the paper's md or bib files, type:

```
@whedon generate pdf
```

**Comment by whedon** [comment]
```
Software report (experimental):

github.com/AlDanial/cloc v 1.88  T=0.20 s (107.1 files/s, 130818.9 lines/s)
-------------------------------------------------------------------------------
Language                     files          blank        comment           code
-------------------------------------------------------------------------------
Python                           5           1785           2970          13262
Bourne Shell                     2            653            500           3569
Markdown                         8            454              0           1745
TeX                              1             55              0            559
YAML                             4              5              0             55
INI                              1              8              0             32
-------------------------------------------------------------------------------
SUM:                            21           2960           3470          19222
-------------------------------------------------------------------------------


Statistical information for the repository '8c1253e92c1faba8877e0193' was
gathered on 2021/05/19.
The following historical commit information, by author, was found:

Author                     Commits    Insertions      Deletions    % of changes
Francesco Bruni                  1             7              0            0.01
Frankie Robertson                1             1              1            0.00
Gregory Boddin                   2             4              2            0.01
Jonathan Passerat-Pa             1             3              0            0.01
Luís Alves                      20          1500            259            3.66
Manabu ISHII                     1             2              1            0.01
Michael R. Crusoe                2             7              4            0.02
Pau Ruiz Safont                  1             1              1            0.00
Romain Reuillon                  1             3              0            0.01
jorge                            1          2725            882            7.50
mariojmdavid                    38          7279           6615           28.90
udocker support                206         21354           7419           59.86

Below are the number of rows from each author that have survived and are still
intact in the current revision:

Author                     Rows      Stability          Age       % in comments
Francesco Bruni               3           42.9         35.1                0.00
Gregory Boddin                1           25.0         29.6                0.00
Jorge Gomes                 960          100.0         49.6                8.02
Luís Alves                  485           32.3         40.6                2.06
Manabu ISHII                  1           50.0         34.7                0.00
Michael R. Crusoe             6           85.7         31.1                0.00
Pau Ruiz Safont               1          100.0         35.3              100.00
Romain Reuillon               1           33.3         38.8                0.00
mariojmdavid               5728           78.7         14.6               26.96
udocker support           10831           50.7         29.8                7.25
```

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03295/joss.03295/10.21105.joss.03295.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03295/joss.03295/10.21105.joss.03295.pdf) :page_facing_up: :point_left:

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1016/j.cpc.2018.05.021 is OK
- 10.6084/m9.figshare.3115156.v2 is OK
- 10.1038/s41592-018-0046-7 is OK
- 10.1016/j.future.2013.05.003 is OK
- 10.1016/j.future.2018.01.022 is OK
- 10.1007/s10723-018-9453-3 is OK
- 10.1140/epjc/s10052-019-7382-3 is OK
- 10.1093/gigascience/giz022 is OK
- 10.1186/s13321-020-0408-x is OK
- 10.1109/ACCESS.2020.2964386 is OK
- 10.1093/gigascience/giz014 is OK
- 10.1016/j.envsoft.2019.104585 is OK
- 10.1016/j.future.2013.05.003 is OK
- 10.1016/j.jpdc.2019.08.002 is OK
- 10.3204/PUBDB-2018-00782/B8 is OK
- 10.1007/978-3-030-34356-9_36 is OK
- 10.1093/nar/gkaa1125 is OK
- 10.1051/epjconf/202024503002 is OK
- 10.1186/s13321-020-0408-x is OK
- 10.1007/s10723-021-09543-5 is OK
- 10.3390/app11041438 is OK
- 10.1051/epjconf/202024507032 is OK
- 10.5194/egusphere-egu21-8418 is OK
- 10.1088/1742-6596/898/8/082039 is OK
- 10.1016/j.procs.2017.05.065 is OK
- 10.1016/j.procs.2016.05.362 is OK
- 10.1371/journal.pone.0177459 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by danielskatz** [comment]
@vsoch and @mviereck - Thanks for agreeing to review this submission.
This is the review thread for the paper. All of our communications will happen here from now on.

Both reviewers have checklists at the top of this thread with the JOSS requirements. As you go over the submission, please check any items that you feel have been satisfied. There are also links to the JOSS reviewer guidelines.

Please read the first couple of comments in this issue carefully, so that you can accept the invitation from JOSS and be able to check items, and so that you don't get overwhelmed with notifications from other activities in JOSS.

The JOSS review is different from most other journals. Our goal is to work with the authors to help them meet our criteria instead of merely passing judgment on the submission. As such, reviewers are encouraged to submit issues and pull requests on the software repository. When doing so, please mention `openjournals/joss-reviews#3295` so that a link is created to this thread (and I can keep an eye on what is happening). Please also feel free to comment and ask questions on this thread. In my experience, it is better to post comments/questions/suggestions as you come across them instead of waiting until you've reviewed the entire package.

We aim for reviews to be completed within about 2-4 weeks, ideally. Please let me know if either of you require some more time. We can also use Whedon (our bot) to set automatic reminders if you know you'll be away for a known period of time.

Please feel free to ping me (@danielskatz) if you have any questions/concerns.

**Comment by danielskatz** [comment]
👋 @jorge-lip - I just want to point out that there are now some issues that have arisen in the review process...

**Comment by mviereck** [comment]
@jorge-lip Could you please give some feedback, here or in the tickets? I've seen you are back contributing to udocker again.

@mariojmdavid You are an active contributer of udocker. Maybe you could have a look here, too?

If you don't have the time yet to work at the review, please tell us.

Maybe it makes sense to finish the port to python3 first (and afterwards wait some time for possible bug reports by users) before continuing the review (compare https://github.com/indigo-dc/udocker/issues/313). A beta version does not fit the requirements of joss, and the python2 version is going to be outdated.

I assume it would be no problem to pause the review until the work is done.

**Comment by mariojmdavid** [comment]
we have been busy preparing a new release which we plan over the next few days, this is the version in devel3 branch that we consider mature for release and supports python3. We have been also reviewing the documentation and its location, as well as ensuring the proper run of the Jenkins CI pipeline with success

**Comment by danielskatz** [comment]
We should decide what's being reviewed here - it's fine for changes to be made in response to review comments, but if there's going to be a new version, perhaps we should not continue this review and start a new review when that version is ready.  Otherwise, it's unfair to the reviewers.  Alternatively, the submitted version (plus changes made in response to the review) can continue to be reviewed.

**Comment by jorge-lip** [comment]
Hi,
udocker supports both the Python 2 only version as 1.1.x series and the Python 3 version as 1.2.x series. We will continue maintaining both in parallel for some more time to allow a smooth migration. Both versions reside in different branches and have been maintained in parallel, they are equal in terms of features and user interface, the documentation has very small differences. The differences are mainly in the code structure.

For the sake of clarity we would like to move the Python 3 version to the master branch as this also makes more sense. I hope moving these branches addresses the concerns raised regarding Python 2 and 3 in the best way for everybody. The changes that are ongoing are the adjustments  needed for doing this move, and we are also adding the corrections and improvements suggested by the reviewers to both versions.

Regarding the question on tests we always had SQA tests. udocker must pass these tests before each release and this applies to both versions. We also use Jenkins and have a pipeline that executes the unit tests, integration tests and bandit for security.

**Comment by vsoch** [comment]
> They are equal in terms of features and user interface, the documentation has very small differences. The differences are mainly in the code structure.

If the two versions are equal in terms of all of these things and it's primarily the structure of the code that has changed, I'd like to ask again why this warrants another publication? An update from Python 2.x to 3.x with function and feature equivalent I am not convinced is sufficient enough for a new publication.

**Comment by jorge-lip** [comment]
As I explained there is a lot of changes in udocker from the version that was published initially back in 2018 to the one we have today, there have been several releases since then. These changes have been applied to both versions Python 2 and Python 3. The changes are in the changelog.

**Comment by vsoch** [comment]
Gotcha, thanks @jorge-lip. I'd like to request that those changes are highlighted in the paper, and it also makes sense to reference your first paper so the reader has context to go from.

**Comment by whedon** [comment]
:wave: @mviereck, please update us on how your review is going (this is an automated reminder).

**Comment by whedon** [comment]
:wave: @vsoch, please update us on how your review is going (this is an automated reminder).

**Comment by mviereck** [comment]
@whedon Just waiting for progress in the tickets. Also it seems that we will review the yet unreleased python3 version. So I've paused reviewing.

**Comment by whedon** [comment]
I'm sorry human, I don't understand that. You can see what commands I support by typing:

```
@whedon commands
```

**Comment by vsoch** [comment]
[at]whedon, I did finish my preliminary review but I don't think there are responses to my issues yet. And +1 what @mviereck said, it's not clear what we are reviewing, and I'm also wanting more focus on how this version is different from the previous publication. I'm not sure I'm convinced that the changes are substantial enough to warrant a new paper.

**Comment by mviereck** [comment]
Just want to note that the third party tools of udocker are not clearly stated and are missing source, license and origin. This needs to be documented well. In the current state this is an acceptance blocker. Compare https://github.com/indigo-dc/udocker/issues/311

**Comment by jorge-lip** [comment]
@whedon generate pdf

**Comment by jorge-lip** [comment]
The changes since v1.1.1 are now highlighted in text. The external tools used by udocker are highlighted in the documentation and in the acknowledgements, the licenses are included in the tools tarball and since the just released v1.3.0  they are extracted to the installation directory under (.udocker/doc). The installation documentation was revised and now makes clear what source code repositories are used to build the tools. The script to build the tools was already provided in the udocker repository in utils/build_tarball.sh

**Comment by danielskatz** [comment]
Thanks - I'll look forward to hearing opinions from @vsoch and @mviereck

**Comment by mviereck** [comment]
There is already a [paper for udocker](https://www.sciencedirect.com/science/article/pii/S0010465518302042) as @vsoch already pointed out in the preview.
udocker has seen some development since then, but the major difference is still not obvious.
Discussion at: https://github.com/indigo-dc/udocker/issues/312

@danielskatz It is normal that software evolves further after a paper has published. How is this handled within joss? I did not found a guidance for this.
I found a rule that software should have reached a stable state before publishing a paper and should not expect major or even breaking changes.
But how is regular development handled? What if someone cites a tool but uses the recent version instead of the one published in the paper?
I assume those points are already discussed somewhere. This might help here.

**Comment by danielskatz** [comment]
The general feeling on this in JOSS is that each JOSS paper should be a clear contribution, in the same way that a research project can lead to multiple papers, some of which can be incremental progress in an area that was previously published. It's up the reviewers to decide if there is sufficient new material/progress.  This is described in https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort

Another view of this is that it might be reasonable to have a new paper for a new major release, but typically isn't for a minor release, assuming major releases don't happen too frequently.

**Comment by jorge-lip** [comment]
Thanks for the feedback we are improving the description of the developments performed since the last publication, we hope provide a more elaborated description tomorrow.

**Comment by jorge-lip** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03295/joss.03295/10.21105.joss.03295.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03295/joss.03295/10.21105.joss.03295.pdf) :page_facing_up: :point_left:

**Comment by jorge-lip** [comment]
A more extensive section on the developments since the last publication is now available.
The repositories and licenses for the tools have been added to the installation manual on section 6 as requested.
https://indigo-dc.github.io/udocker/installation_manual.html

**Comment by danielskatz** [comment]
Thanks @jorge-lip - let's see what @vsoch and @mviereck think now

**Comment by vsoch** [comment]
I'm good! My concerns with respect to the checklist were:

 - no information about automated tests
 - missing a proper discussion about other container technologies / comparisons
 - installation instructions

And the authors have resolved these issues so my checklist is complete.

**Comment by danielskatz** [comment]
Thanks @vsoch!

**Comment by mviereck** [comment]
The installtion instructions and dependency declaration has been rewritten and clearly state the third party tools now including their source, license and origin. This is done well and no longer an acceptance blocker.

Now I continue to work on the checklist.

**Comment by danielskatz** [comment]
👋 @mviereck - How is this going? Is anything blocking your progress?

**Comment by mviereck** [comment]
Sorry that I've been away that long. I was distracted by RL.
Most remaining points I just have to check without expecting a problem.

One major point was a missing description of the difference to the previous paper.
Now there is a long chapter on this. It seems to me that this fits in the way you've said:
> in the same way that a research project can lead to multiple papers, some of which can be incremental progress in an area that was previously published.

Overall the paper is quite long and detailed. A bit I am missing is to bring it to a short and clear point. I asked for this in https://github.com/indigo-dc/udocker/issues/312 but got no response.

A bit I am lost in getting the full content of the paper just because it is a bit hard to me to grok it in a non-native language.

To bring it to a point: The paper is rather written like a full documentation than like a core desciption that comes to a point. This can be weighted good or bad. A scientific paper on research should show a lot of detail. But I'd say a paper on software should tell less detail but should rather show the essential core.

**Comment by jorge-lip** [comment]
@danielskatz Along these weeks we have put a lot of effort to address the recommendations and requests both in terms of the software itself and also in terms of improving the paper that indeed now starts to become very close to a full paper. At this point all the material and detailed information to decide if this work is suitable for publication is available. We are happy to perform further improvements if needed but first we need to know if the work is to be published. I wish to thank the reviewers for their positive suggestions and contributions to improve udocker.

**Comment by danielskatz** [comment]
@jorge-lip - the expectation for JOSS papers is that they will be improved via collaboration between the reviewers and the authors to the point where they can be published, unless there is an issue of scope, which I don't think it the case here.

On the other hand, a typical paper is fairly short, with most of the description of the software in the documentation (README, etc.), not in the paper. I think we have gotten away from that goal, but perhaps this is ok for this software.

**Comment by jorge-lip** [comment]
Many thanks for the clarifications, therefore assuming that the paper is in scope and since all software and documentation issues have been addressed, just let us know in this thread if there are additional points that you feel must still be corrected in the paper.

**Comment by danielskatz** [comment]
👋 @mviereck - To try to wrap this up, can you be specific about the changes you want to see to finish checking off the items in your review list?  Perhaps as bullets or as new issues in the source repo?

Of course, this is in addition to https://github.com/indigo-dc/udocker/issues/312, where it looks like @jorge-lip still needs to respond.

**Comment by mviereck** [comment]
@danielskatz  Basically this one question brings it to the point:
>  If someone writes a scientific paper and wants to cite udocker: at which point he would have the urgent need to cite the new paper instead of the previous one?

As long as this cannot be answered, there is no obvious need for a new paper and I cannot check of "substantial scholary effort". 

Other remaining points in the checklist I just  have to work through. If I find issues there, I'll open new tickets.

**Comment by jorge-lip** [comment]
> If someone writes a scientific paper and wants to cite udocker: at which point he would have the urgent need to cite the new paper instead of the previous one?

From our point of view yes version 1.3.0 is largely different from 1.1.0. These differences and new developments are described in the paper in section "Developments since 1.1.1".

**Comment by danielskatz** [comment]
>> If someone writes a scientific paper and wants to cite udocker: at which point he would have the urgent need to cite the new paper instead of the previous one?

> As long as this cannot be answered, there is no obvious need for a new paper and I cannot check of "substantial scholary effort".

Please note that the [discussion about this in the JOSS documentation](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort) says:

> As a rule of thumb, JOSS’ minimum allowable contribution should represent not less than three months of work for an individual. Some factors that may be considered by editors and reviewers when judging effort include:
> 
> * Age of software (is this a well-established software project) / length of commit history.
> * Number of commits.
> * Number of authors.
> * Total lines of code (LOC). Submissions under 1000 LOC will usually be flagged, those under 300 LOC will be desk rejected.
> * Whether the software has already been cited in academic papers.
> * Whether the software is sufficiently useful that it is likely to be cited by your peer group.
> 
> In addition, JOSS requires that software should be feature-complete (i.e. no half-baked solutions) and designed for maintainable extension (not one-off modifications of existing tools). “Minor utility” packages, including “thin” API clients, and single-function packages are not acceptable.

Along with the comment from @jorge-lip in the comment above this one.  I think the work since the previous paper meets these factors, but if you don't, let's talk about which factors you think are the problem. (And note that this is not a case where all factors must be met - these are just samples of factors that could be considered.)

**Comment by danielskatz** [comment]
👋 @mviereck - How are things going with the remaining items on your checklist?

**Comment by mviereck** [comment]
I was still hoping for an answer in https://github.com/indigo-dc/udocker/issues/312 .
Yet I am working on a comparision to answer the questions myself.

**Comment by danielskatz** [comment]
Thanks @mviereck.

👋 @jorge-lip, How would you answer the questions in this issue?

**Comment by danielskatz** [comment]
👋 @jorge-lip - how are you doing in responding to the issues from the @mviereck?  Can we get an update here?

**Comment by mviereck** [comment]
@jorge-lip provided a more [comprehensive summary of the changes](https://github.com/indigo-dc/udocker/issues/312#issuecomment-875747957) done in udocker. It boils down to good code maintenance and a broader support of different environments.
The essential idea and concept is still the same and fits the description in the first paper. So there is basically no need for a new paper except to provide a recent version number.

Though, the changes are big enough to fit the definition of "incremental progress":
> The general feeling on this in JOSS is that each JOSS paper should be a clear contribution, in the same way that a research project can lead to multiple papers, some of which can be incremental progress in an area that was previously published.

However, my feeling is that JOSS needs a better handling of software versioning. A well  done paper describes the idea and concept of a software without going too much into details like API description or supported environments. 
Good maintenance can be a lot of work, while the paper still fits the software after years. This is the case here. 
A solution to update software versions to existing papers in JOSS is desirable, so one could cite a paper referring to recent versions.

The first paper hasn't been published at JOSS. So the review still has its use for the documentation of udocker that has severely improved meanwhile.

---------------

Currently I wait for an answer in ticket https://github.com/indigo-dc/udocker/issues/348 . The paper lists a lot of authors that do not appear as contributors on github.

**Comment by danielskatz** [comment]
@mviereck - I think the changes/work are sufficient for a new version of the paper, so let's take this as ok from the point of view of the journal (though I agree it's a complicated issue where we don't have a good simply-stated policy), and focus on other issues, like the question about the authors. In addition to this, I see three more unchecked boxes in your review - what information/changes would be needed to check these?

**Comment by mviereck** [comment]
> I think the changes/work are sufficient for a new version of the paper, so let's take this as ok from the point of view of the journal 

I have checked this one off.

> In addition to this, I see three more unchecked boxes in your review - what information/changes would be needed to check these?

I am just checking point after point. Once I've checked one off, I go to the next. So I did not look closer yet to the remaining points.

**Comment by danielskatz** [comment]
@mviereck - just checking on your progress...

**Comment by mviereck** [comment]
@danielskatz I am still waiting for a response in https://github.com/indigo-dc/udocker/issues/348.

**Comment by danielskatz** [comment]
@jorge-lip - can you respond to this issue?

**Comment by danielskatz** [comment]
👋 @jorge-lip - can you respond to this issue?

**Comment by mviereck** [comment]
Finally, about 4 weeks later, I got a short response by @jorge-lip  about the list of authors.
The list might be appropriate or not, I don't know. It contains beta testers and others, too.

It is very hard to get any response since the very begin of the review.
I don't want to ask any questions anymore and than to wait and beg for answers. There are more reminders then responses. 
I've lost my good will to get the paper through the review.

I vote to reject the paper due to missing cooperation by the submitting author.

**Comment by danielskatz** [comment]
@mviereck - I think at this point the best thing to do is to remove you and for me to find another reviewer. I agree that this has been a slow process, but waiting for each concern to be addressed before moving to the next one has contributed to this being slow.

So thank you for your contributions to the review process, and don't worry it any further - we'll carry on without you from here.

**Comment by danielskatz** [comment]
@whedon remove @mviereck as reviewer

**Comment by whedon** [comment]
OK, @mviereck is no longer a reviewer

**Comment by danielskatz** [comment]
@vsoch - do you by any change have any suggestions for someone who could in and review this?  Given that you and the author have worked through a number of issues, I hope there won't be any/too many more to resolve.

**Comment by vsoch** [comment]
Possibly @reidpr or someone from his team that works on CharlieCloud - there aren’t that many folks out there with deep container expertise.

**Comment by reidpr** [comment]
Given that I lead Charliecloud which is or at least could be perceived to be a competing product, I probably have a conflict of interest.

Unrelated, authors might find our pre-print interesting. It covers the workings unprivileged containers in some detail and offers a new classification scheme for container privilege models. It will be published at SC21 this fall. https://arxiv.org/abs/2104.07508

**Comment by danielskatz** [comment]
@reidpr - Working on another product in this space isn't a conflict, if you think you could review this fairly. Expertise is a good thing to bring to a review

**Comment by vsoch** [comment]
@danielskatz both reviewers have expressed concerns about publishing this paper - and it seems like they aren't being heard -  @mviereck put a lot of time/energy into the review and wasn't satisfied at the end, and instead of hearing his concerns (which are somewhat overlapping with mine) he's simply removed from the process. What is the point of having reviewers if you aren't listening to them? :thinking:

**Comment by danielskatz** [comment]
@vsoch - Did you feel that your concerns weren't addressed?  I know you had some, but I thought from your agreeing that this could be published after discussions with me and the author, you were ok with this, given your comments in https://github.com/openjournals/joss-reviews/issues/3295#issuecomment-863345897 in particular.

The issue with @mviereck and the author is something different to me, where the relationship broke down - note that the problems in https://github.com/openjournals/joss-reviews/issues/3295#issuecomment-894837288 are generally not technical but are based on this problematic relationship.  Given this, it seemed to me that a change was needed.

**Comment by vsoch** [comment]
That was my response after I expressed concerns to you privately and you emailed me back that you "don't have any issue with accepting this" and I shrugged my shoulders and decided that the other reviewer would be the deciding factor. But now that the other reviewer has come to the same conclusions as me (and a few more) I'm thinking that my views weren't so unfair or far off. To reiterate my concerns, I believe that a refactor to Python 3 and adding a few new features does not constitute enough for a new publication, especially since the first was only in 2018. There are so many pieces of software out there that have updated versions, to Python 3, and they don't pursue publication because the core of the software just isn't different. This entire review feels really strange to me, almost like what we have to say doesn't matter.

**Comment by vsoch** [comment]
I think I would also like to be removed as a reviewer, thank you.

**Comment by danielskatz** [comment]
I'm now asking other JOSS AEiCs for advice on how to handle this - perhaps my judgement that this was in scope was incorrect, and we should have had all the JOSS editors vote.

**Comment by mariojmdavid** [comment]
dear reviewers and editor,
it is unfortunate that things have come to this point of rupture.
the first question is/was of course if this new version was appropriate to be published in the JOSS, this has been questioned in the very beginning of the process, and it seemed the answer was yes, and thus the effort from absolutely everybody to improve, the paper itself, but as well the SW and the documentation, and from my point of view, indeed the comments and scrutiny from the reviewers where precious for such improvements.

the release of the 1.3.0 version was somewhat disruptive but necessary, since it was being improved and tested in real production environments by several users for many months,  and that is why it was deemed mature for production release, while it made all the sense that the paper would be about this release.

I have been into both sides of the process of review of papers, as well as of review of projects, in either case I have seen my papers and my projects being rejected, as well as I have rejected both papers and projects.
when I reject a paper or a project I say outright from the beginning, and of course state all the reasons to do that
it does not mean that I may not change my mind or vote later with discussion with other reviewers that have other or different viewpoints that I may come to agree with.

Also happens that I accept a paper but with comments that should be implemented in order to be published, I think this has to be evaluated from the beginning, if a paper has the potential to be publish but needs improvement, or if the paper is not appropriate for publishing, you say it outright in the beginning.

As I stated earlier, whatever the decision is to publish or not the paper, the SW and the documentation have had a large improvement, and of course this was due to the work of the reviewers, and ours on the implementation.

with kind regards
Mario David

**Comment by vsoch** [comment]
Thank you @mariojmdavid. Just to be transparent, my concerns were there from the beginning, and I emailed the editor here and was sort of pushed in the direction of "I think this is okay" (subtext you should too) and I proceeded with the review. I didn't feel that I had the power or right to be assertive about my opinion, so I stopped there. I think if I were truly in charge I would have concluded (before any review) that I didn't consider this second paper appropriate for a piece of software that already exists and has been published - in my mind updates to or new features, refactor to different structure or language, does not constitute enough for a paper. And I proceeded rather submissively thinking I didn't think I really had a choice to outright reject reviewing it. 

But regardless of my critique - I realize that you are a person with priorities, feelings, etc., and I'm really sorry for not being more up front about this. This says nothing about my feelings about udocker (I'm a huge fan of container technologies, and I champion your work) but rather if I think it's appropriate for a publication. I also doubt myself in this entire process that maybe my standards are too high and I should just give the a-ok, but I've had unpublished software rejected for (what I deem to be) much sillier reasons, so I don't think my intuition is that unreasonable. Part of this practice / exercise is convincing the reviewers that there is some paramount change that warrants a second publication, and I was not convinced.

**Comment by mariojmdavid** [comment]
@vsoch thanks for your directness and I totally understand, of course I as one of the authors and contributors would like to see this published, but of course it's up to the editors and reviewers to evaluate that, and I will accept whatever decision is made, and I say sincerely on my part that even if it is not accepted for publication, it was a productive exercise on improvements,
I would like to say that I am expressing my point of view and that does not reflect necessarily the same views from the other authors

**Comment by danielskatz** [comment]
@vsoch - I don't want to make this a major point, but I've generally tried to keep all conversations as public in the issue, so that there are no questions about who said what. I wish we had done this in this case, rather than our discussion happening in another channel.

While I accept that you are saying what you believe happened, this was not my perception or intent at the time, which was much more that there were questions and I was providing an opinion by talking about JOSS guidelines, though in hindsight, I recognize that my JOSS position gave my opinion more weight than I intended.  I took your acceptance to mean you were ok with the decision that I thought we collectively agreed on, which I now realize was not correct.

I apologize for not being more aware of all your and @mviereck's opinions, and for doing something that you felt pressured you to act in a way you didn't want to act, and am glad you brought this up today. I will be much more careful about things like this in the future.

**Comment by danielskatz** [comment]
The JOSS AEiCs have now looked at this review thread and the software, and I'm going to let @arfon as the overall JOSS AEiC take over to state the decision as a person not directly involved in the process.

**Comment by reidpr** [comment]
As someone (possibly) invited to review, I'm not familiar with this journal nor its review process.

But, seeing R1 (@mviereck) removed from the paper immediately after their reject recommendation, and R2 (@vsoch) feeling her concerns aren't being heard, makes it clear to me I do not want to be involved. Thanks for thinking of me and good luck.

**Comment by danielskatz** [comment]
@reidpr - this has been a very unusual review, so please don't take this as typical of the journal or most submissions or reviews - see https://github.com/openjournals/joss-reviews/issues/3457 as one that was much more straightforward and typical, with a happy author, and https://github.com/openjournals/joss-reviews/issues/3311 as another, from @vsoch.  Unfortunately, sometimes (and I think rarely) things don't go as well, but we can learn from these cases and do better.

**Comment by arfon** [comment]
This has been a challenging review for many reasons and I would like to thank @vsoch and @mviereck for their efforts and time invested here and the authors for their work to improve their submission based on the feedback of the reviewers. 

@jorge-lip @vsoch @mviereck – as @danielskatz mentioned, the broader EiC team at JOSS has reviewed the status of this submission and arrived at a decision about how to proceed. 

After reviewing the feedback from @mviereck & @vsoch and conducting our own internal review, we agree with their assessment that the work done between the [earlier paper](https://www.sciencedirect.com/science/article/pii/S0010465518302042) and this submission is insufficient to warrant a second publication. As such, this submission doesn't meet JOSS' [substantial scholarly effort](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort) criterion.

**Comment by arfon** [comment]
@whedon reject

**Comment by whedon** [comment]
Paper rejected.
